pkg/ccl/changefeedccl/checkpoint/checkpoint_test.go

```diff
@@ -164,6 +166,166 @@ func TestCheckpointMake(t *testing.T) {
 	}
 }
 
+func TestCheckpointRestore(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ts := func(wt int64) hlc.Timestamp {
+		return hlc.Timestamp{WallTime: wt}
+	}
+
+	for name, tc := range map[string]struct {
+		initialHighWater       hlc.Timestamp
+		oldCheckpointTs        hlc.Timestamp
+		oldCheckpointSpans     []roachpb.Span
+		newCheckpoint          jobspb.TimestampSpansMap
+		expectedError          error
+		expectedFrontierForOld checkpointSpans
+		expectedFrontierForNew checkpointSpans
+	}{
+		"old checkpoint ts is empty": {
+			initialHighWater: ts(0),
+			oldCheckpointTs:  ts(0),
+			oldCheckpointSpans: []roachpb.Span{
+				{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")},
+			},
+			newCheckpoint: jobspb.TimestampSpansMap{
+				Entries: []jobspb.TimestampSpansMap_Entry{
+					{
+						Timestamp: ts(2),
+						Spans:     []roachpb.Span{{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")}},
+					},
+				},
+			},
+			expectedError: errors.New("checkpoint timestamp is empty"),
+		},
+		"new checkpoint ts is empty": {
+			initialHighWater: ts(0),
+			oldCheckpointTs:  ts(2),
+			oldCheckpointSpans: []roachpb.Span{
+				{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")},
+			},
+			newCheckpoint: jobspb.TimestampSpansMap{
+				Entries: []jobspb.TimestampSpansMap_Entry{
+					{
+						Timestamp: ts(2),
+						Spans:     []roachpb.Span{{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")}},
+					},
+					{
+						Timestamp: ts(0),
+						Spans:     []roachpb.Span{{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")}},
+					},
+				},
+			},
+			expectedError: errors.New("checkpoint timestamp is empty"),
+		},
+		"same checkpoint over ts": {
+			initialHighWater: ts(0),
+			oldCheckpointTs:  ts(1),
+			oldCheckpointSpans: []roachpb.Span{
+				{Key: roachpb.Key("a"), EndKey: roachpb.Key("e")},
+			},
+			newCheckpoint: jobspb.TimestampSpansMap{
+				Entries: []jobspb.TimestampSpansMap_Entry{
+					{
+						Timestamp: ts(2),
+						Spans: []roachpb.Span{{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")},
+							{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")}},
+					},
+					{
+						Timestamp: ts(1),
+						Spans:     []roachpb.Span{{Key: roachpb.Key("d"), EndKey: roachpb.Key("e")}},
+					},
+				},
+			},
+			expectedFrontierForOld: checkpointSpans{
+				{span: roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("e")}, ts: ts(1)},
+			},
+			expectedFrontierForNew: checkpointSpans{
+				{span: roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")}, ts: ts(2)},
+				{span: roachpb.Span{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")}, ts: ts(2)},
+				{span: roachpb.Span{Key: roachpb.Key("d"), EndKey: roachpb.Key("e")}, ts: ts(1)},
+			},
+			expectedError: nil,
+		},
+		"same checkpoint over ts with non-zero initial highwater": {
```

Comment:
Could you also include a test case where some part of the tracked span space is not checkpointed?

Reply:
We discussed this offline and I missed that the entire span space is `[a,z)`. Might be worth making a single variable for the key space (i.e. `[]roachpb.Span{{Key: roachpb.Key("a"), EndKey: roachpb.Key("z")}`) that goes at the top of the test closure.